src/Chainweb/WebPactExecutionService.hs

```diff
@@ -37,13 +44,13 @@ mkWebPactExecutionService hm = WebPactExecutionService $ PactExecutionService
           Just p -> act p
           Nothing -> throwM (userError $ "PactExecutionService: Invalid chain ID in header: " ++ show h)
 
-mkPactExecutionService :: TQueue RequestMsg -> PactExecutionService
-mkPactExecutionService q = PactExecutionService
+mkPactExecutionService :: MempoolBackend ChainwebTransaction -> TQueue RequestMsg -> PactExecutionService
+mkPactExecutionService mempool q = PactExecutionService
   { _pactValidateBlock = \h pd -> do
       mv <- validateBlock h pd q
       r <- takeMVar mv
       case r of
-        Right pdo -> return pdo
+        Right pdo -> markAllConfirmed mempool pdo >> return pdo
```

Comment:
So in case of a fork, the user would have to resubmit? Forks at depth 0 are not that rare. I guess, it is good enough for testnet?

Reply:
It is good enough for test net. Will fix reintroduction tomorrow

Reply:
Would it be possible to change the is to something like `markAllConfirmedForBlock` and tag the confirmed transactions with the BlockHeader. 

That would allow to track when a transaction has to be re-submitted because the block ends up on a fork. It would be relatively straight forward to add the event of a block moving to a fork to the `WebBlockHeaderDB` BlockHeader Stream.

Reply:
Currently `WebBlockHeaderDB` emits a stream of `BlockHeader`. We could change that to a stream of something like `BlockHeaderEvent = NewBlockHeader BlockHeader | ForkBlockHeader BlockHeader`.

Reply:
I don't have much time but I have an algorithm for the mempool update already -- basically mempool has to watch the `Cut` tvar. When it sees an update, it walks both sides of the old and new cut back to the `greatestCommonAncestor`, accumulating tx lists as it goes. You reintroduce all of the tx hashes that were on the losing fork but didn't make it on the winning one.

As a result of doing this, the mempool consensus engine will publish a `TVar (Cut, Map ChainId (Vector ChainwebTransaction))`, and this is what miners will mine on. Does that sound ok to you

Reply:
Consensus is already doing this internally, computing forks. So, it would make sense to just publish this as an "ConsensusEvent" instead of expecting each client to recompute it. It's also useful information for other consumers and not just the mem-pool.

I am still not comfortable with the mempool deciding on which cut to mine on. Deciding the winner is inherently the job of consensus. Mempool should provide transactions for a given parent cut, but not chose the parent cut itself.

Reply:
Currently the flow is that the consensus publishes a new cut. The miner sees the cut and asks pact/mempool to create a block for that particular cut. I think, that's a simple and intuitive flow.

Reply:
It's the cut-processing pipeline that drives the consensus control flow. I think we shouldn't change that.

Reply:
But I see the benefit of allowing the mem-pool to preempt mining if it wants to update the payload for the currently mind block. Let's talk about that later.